servicediscovery/mdns/mdns.go

```diff
@@ -0,0 +1,75 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package mdns
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dapr/components-contrib/servicediscovery"
+	"github.com/grandcat/zeroconf"
+)
+
+func NewMDNSResolver() *Resolver {
+	return &Resolver{}
+}
+
+type Resolver struct {
+	once     sync.Once
+	resolver *zeroconf.Resolver
+}
+
+func (z *Resolver) ResolveID(req *servicediscovery.ResolveRequest) (string, error) {
+	port, err := z.LookupPortMDNS(req.ID)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("localhost:%v", port), nil
+}
+
+// LookupPortMDNS uses mdns to find the port of a given service entry on a local network
+func (z *Resolver) LookupPortMDNS(id string) (int, error) {
+	var err error
+	var t *zeroconf.Resolver
+	z.once.Do(func() {
+		t, err = zeroconf.NewResolver(nil)
+		z.resolver = t
+	})
```

Comment:
What about this scneario? 

if NewResolver throws the error at the first call, then this block will never be called by the next call. then err is not caught in line 43-46 then it will make panic.

Reply:
Refactored